### PR TITLE
feat(udf): minimal Python UDF SDK

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR designs a minimal SDK for Python UDFs.

Now you can define a function in Python like this:

```python
from risingwave.udf import udf, UdfServer

@udf(input_types=['INT', 'INT'], result_type='INT')
 def gcd(x: int, y: int) -> int:
     while y != 0:
         (x, y) = (y, x % y)
     return x

if __name__ == '__main__':
     server = UdfServer()
     server.add_function(gcd)
     server.serve()
```

This PR also fixes the problem when functions have no input arguments.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] ~~I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).~~
- [x] ~~I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### CODE REVIEW

File: src/udf/arrow_flight.py [hunk 1]
```diff
@@ -1,55 +0,0 @@
-import pathlib
```
Comment (xxchan) [inline src/udf/arrow_flight.py:1]: remember to update `src/udf/README.md`

File: e2e_test/udf/test.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import sys
+sys.path.append('src/udf/python')  # noqa
```
Comment (xxchan) [inline e2e_test/udf/test.py:2]: What is `noqa` 🥵

Comment (xxchan) [reply e2e_test/udf/test.py:2]: > Adding # noqa to a line indicates that the linter (a program that automatically checks code quality) should not check this line. Any warnings that code may have generated will be ignored.

WOW 😲

File: ci/scripts/run-e2e-test.sh [hunk 3]
```diff
@@ -55,6 +55,11 @@ sqllogictest -p 4566 -d dev './e2e_test/batch/**/*.slt' --junit "batch-${profile
 sqllogictest -p 4566 -d dev './e2e_test/database/prepare.slt'
 sqllogictest -p 4566 -d test './e2e_test/database/test.slt'
 
+echo "--- e2e, ci-3cn-1fe, udf"
+python3 e2e_test/udf/test.py &
```
Comment (xxchan) [inline ci/scripts/run-e2e-test.sh:59]: Do we have python in CI? 🥵

File: e2e_test/udf/python.slt [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Before running this test:
+#   python3 e2e_test/udf/test.py
+
+# TODO: check the service on creation
```
Comment (xxchan) [inline e2e_test/udf/python.slt:4]:
```suggestion
# TODO: check the service on creation. 
# Currently whether the function exists and whether the signature matches is not checked.
```
😄

File: e2e_test/udf/python.slt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+create function int_42() returns int as 'http://localhost:8815' language arrow_flight;
+
+statement ok
+create function gcd(int, int) returns int as 'http://localhost:8815' language arrow_flight;
+
+# Create a function with the same name but different arguments.
+statement ok
+create function gcd(int, int, int) returns int as 'http://localhost:8815' language arrow_flight;
+
+# Create a function with the same name and arguments.
```
Comment (xxchan) [inline e2e_test/udf/python.slt:17]: What about creating a function with the same name as a build-in function? 🥵

File: src/udf/python/risingwave/udf.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def _create_udf(f, input_types, result_type, name):
+    return UserDefinedFunctionWrapper(
+        f, input_types, result_type, name)
+
+
+def udf(input_types: Union[List[pa.DataType], pa.DataType],
+        result_type: pa.DataType,
+        name: Optional[str] = None,) -> Union[Callable, UserDefinedFunction]:
```
Comment (xxchan) [inline src/udf/python/risingwave/udf.py:81]:
```suggestion
def udf(input_types: Union[List[Union[str, pa.DataType]], Union[str, pa.DataType]],
        result_type: Union[str, pa.DataType],
        name: Optional[str] = None,) -> Union[Callable, UserDefinedFunction]:
```
😇

File: src/udf/python/risingwave/udf.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    def add_function(self, udf: UserDefinedFunction):
+        """Add a function to the server."""
+        name = udf.full_name()
+        if name in self._functions:
+            raise ValueError('Function already exists: ' + name)
+        print('added function:', name)
+        self._functions[name] = udf
+
+    def do_exchange(self, context, descriptor, reader, writer):
+        """Run a simple echo server."""
```
Comment (xxchan) [inline src/udf/python/risingwave/udf.py:114]: fix me?

**Review by xxchan — APPROVED** [review]
lgtm

File: src/udf/src/lib.rs [hunk 8]
```diff
@@ -38,11 +38,15 @@ impl ArrowFlightUdfClient {
...
-        for arg in &args.fields {
-            path.push(format!("{}", arg.data_type()));
+        // path = name/[args,]*
+        let mut path = name.to_string() + "/";
+        for (i, arg) in args.fields.iter().enumerate() {
+            if i != 0 {
+                path += ",";
+            }
+            path += &arg.data_type().to_string().to_lowercase();
         }
```
Comment (BugenZhao) [inline src/udf/src/lib.rs:48]: What about extracting this snippet to make the naming consistent everywhere? Like...
https://github.com/risingwavelabs/risingwave/blob/896d92ee3cad6095e021f74b0686777f3d5a33c6/src/frontend/src/handler/create_function.rs#L86-L97

**Review by BugenZhao — APPROVED** [review]
🚀🚀🚀

File: e2e_test/udf/python.slt (hunk 5 above)
Comment (wangrunji0408) [reply e2e_test/udf/python.slt:17]: At present, the creation will succeed but the usage will still refer to the built-in version. I will fix it later. 🥵

File: e2e_test/udf/test.py (hunk 2 above)
Comment (wangrunji0408) [reply e2e_test/udf/test.py:2]: If I don't add this hint, the formatter will reorder this line below all imports, which messes things up. 😇

File: src/udf/src/lib.rs (hunk 8 above)
Comment (wangrunji0408) [reply src/udf/src/lib.rs:48]: Good idea! But this snippet will be deprecated soon. I'm going to decouple the function name between kernel and UDF server in the next PR. The syntax of CREATE FUNCTION will be changed correspondingly to:
```sql
create function gcd(int, int) returns int language python as gcd using link 'http://localhost:8815';
```
where `gcd(int,int)` is the name in kernel and `gcd` is the name in UDF server.

This approach refers to Flink and I think it's better than force unified name. 🤔

File: src/udf/arrow_flight.py (hunk 1 above)
Comment (wangrunji0408) [reply src/udf/arrow_flight.py:1]: Thanks for reminding!

**Pull request merged**